src/mkvgen/src/MkvGenerator.c

```diff
@@ -1606,6 +1606,49 @@ STATUS getSamplingFreqAndChannelFromAacCpd(PBYTE pCpd,
     return retStatus;
 }
 
+STATUS mkvgenGenerateAacCpd(KVS_MPEG4_AUDIO_OBJECT_TYPES objectType, UINT32 samplingFrequency, UINT16 channelConfig, PBYTE pBuffer, PUINT32 pBufferLen)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    INT16 samplingFreqIndex = -1, i, objectTypeInt16 = 0, cpdInt16 = 0;
+
+    CHK(pBufferLen != NULL, STATUS_NULL_ARG);
+    CHK(channelConfig > 0 && channelConfig < MKV_AAC_CHANNEL_CONFIG_MAX && (pBuffer == NULL || *pBufferLen > 0), STATUS_INVALID_ARG);
+
+    for (i = 0; i < MKV_AAC_SAMPLING_FREQUNECY_IDX_MAX; ++i) {
+        if (gMkvAACSamplingFrequencies[i] == samplingFrequency) {
+            samplingFreqIndex = i;
+        }
+    }
+    CHK_ERR(samplingFreqIndex >= 0, STATUS_INVALID_ARG, "Invalid sampling frequency %u", samplingFrequency);
+
+    *pBufferLen = KVS_AAC_CPD_SIZE_BYTE;
+    if (pBuffer == NULL) {
+        CHK(FALSE, retStatus);
+    }
+
+    MEMSET(pBuffer, 0x00, KVS_AAC_CPD_SIZE_BYTE);
+
+    // just in case
+    initializeEndianness();
+
+    /*
+     * aac cpd structure
+     * 5 bits (Audio Object Type) | 4 bits (frequency index) | 4 bits (channel configuration) | 3 bits (not used)
+     */
+    objectTypeInt16 = (INT16) objectType;

```

Comment:
There are two issues with this approach.

1) You are calling getInt16 multiple times and then simply doing 16 bit write.
2) You might be hitting an unaligned address

To solve both of these in one go, work on bytes directly and do bit operations.

Reply:
fix